src/RefImplementations.cc

```diff
@@ -16,11 +16,65 @@
 #include <cmath>
 #include <cstring>
 #include <numeric>
+#include <thread>
 
 using namespace std;
 
 namespace fbgemm {
 
+// Thread-safe random number generator
+//
+// Return a random 32bit integer using xoshiro128++
+// http://prng.di.unimi.it/xoshiro128plusplus.c
+inline uint32_t rnd128_next(int v, int vlen) {
+  constexpr int VLEN_MAX = 16; // max vector size
+  alignas(64) static thread_local uint32_t g_rnd128_buffer[4 * VLEN_MAX];
+  static thread_local bool g_rnd128_initialized = false;
+
+  // Splitmix64: http://prng.di.unimi.it/splitmix64.c
+  auto rnd128_init_next = [](uint64_t &x) {
+    uint64_t z = (x += 0x9e3779b97f4a7c15);
+    z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9;
+    z = (z ^ (z >> 27)) * 0x94d049bb133111eb;
+    return z ^ (z >> 31);
+  };
+
+  auto rotl = [](const uint32_t x, int k) {
+    return (x << k) | (x >> (32 - k));
+  };
+
+  if (!g_rnd128_initialized) {
+    // Initialize rand buffer with uniq values per thread
+    uint64_t h0 = std::hash<std::thread::id>{}(std::this_thread::get_id());
+    for (auto i = 0; i < 4; ++i) {
+      // Use thread hash as seed
+      g_rnd128_buffer[i * VLEN_MAX] = rnd128_init_next(h0);
+      uint64_t h1 = g_rnd128_buffer[i * VLEN_MAX];
+      for (auto v = 1; v < VLEN_MAX; ++v) {
```

Comment:
Can we use a different variable like j? Getting local variable shadowing warning.

Reply:
Sorry this this . Thanks reporting this.